### ISSUE TITLE
Rule: OptionalTypeDeclaration

### ISSUE THREAD

**Issue by vanniktech** [arturbosch/detekt#336]
The following code

```kotlin
private val compositeDisposable: CompositeDisposable = CompositeDisposable()
```

can be simplified by omitting `: CompositeDisposable` since the type can be interfered.

```kotlin
private val compositeDisposable = CompositeDisposable()
```

Obviously this only works when both the type of the declaration and initialization is the same. The following should not be flagged:

```kotlin
private val compositeDisposable: Disposable = CompositeDisposable()
```

since you might only want to work with the `Disposable` type which is implemented by the `CompositeDisposable`. I don't think that we need type resolution for this a simple equals check on the name should be enough.

Any thoughts?

**Comment by Mauin** [comment]
I think the rule makes sense. But disabled by default because I could see lots of project handling optional type declarations on a very case-by-case basis. Sometimes it makes sense for clarity to leave the type on the var/val.

What about inlined functions? Should we also add a rule for those to highlight optional return types?

**Comment by vanniktech** [comment]
Yup obviously off by default. Well maybe the Rule could have options that for functions it's enabled / disabled and the same also for variables.

**Comment by arturbosch** [comment]
I like it coz thats how I judge if the type is necessary. I don't know if on top level declarations the kotlin compiler will complain about missing return type.
